torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -692,42 +722,9 @@ void ProcessGroupNCCL::WorkNCCL::synchronizeInternal(
     }
   }
 
-  // Device synchronize only after we've completed timeout checks.
-  if (barrierTensor_.defined()) {
-    // If we use the work to do barrier, we should block here
-    // `dist.barrier()` only requires all CPU processes to enter this
-    // function, hence we only need to make sure the dummy all-reduce has
-    // completed. So we would only need to sync the **current stream** back to
-    // host, and do not need to synchronize the entire device (which may have
-    // kernels running on other streams).
-    // Using `cudaStreamSynchronize` instead of `cudaDeviceSynchronize` can:
-    // - lower chance of hang;
-    // - CurrentCUDAStream is usually the context of the next operation in
-    // Python, thus blocking current stream would already block the next
-    // compute kernel;
-    // - achieve better barrier performance.
-    auto currentStream = at::cuda::getCurrentCUDAStream(device_.index());
-    // CUDAStream wrapper will correctly use a DeviceGuard here
-    currentStream.synchronize();
-  }
-}
+  // syncrhoize() will block the current stream on the NCCL stream
+  synchronize();
```

Comment:
nit: this change seems to move `synchronizeStream` after blocking wait. 
In 99% of cases, this should be a no-change to code behavior, because the main thread is being hijacked by blocking wait.
I just worry about the 1% chance where another thread my have a handle on the compute stream too -- in case that, it would be nice to call `synchronizeStream` as early as possible.

Reply:
You are right, the comments lie about the 'no behavior change part'. The behavior change here is only when blockingwait mode is true or timeout is specified, where user is more concerned about blocking the CPU thread till the collective is done. Will monitor if any users is affected by this change, Later we might further remove GPU stream synchronize() for blockingwait mode.

Reply:
I am actually not concerned about behavior change, as you mentioned, it only relates to blocking wait mode.
My comment was more on code robustness, as in, would it be generally better to establish stream dependency asap.

Reply:
I actually put synchronize() before the cpu wait code in an early version. But it breaks the current barrier behavior which requires the CPU loop happens before the stream sync (https://www.internalfb.com/code/fbsource/[4713fa33ea08c1d66083e1bbe6c1fa4af785b424]/fbcode/caffe2/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp?lines=695). which I am not sure it is safe to change.  I am planning to remove the barrier blocks in the synchronize() method in future PRs and then we can switch the order